plugins/woocommerce/src/Internal/ReceiptRendering/ReceiptRenderingEngine.php

```diff
@@ -149,15 +161,19 @@ public function generate_receipt( $order, $expiration_date = null, bool $force_n
 			 * - item (only when type is 'product'), and instance of WC_Order_Item
 			 * - quantity (only when type is 'product')
 			 *
-			 * @param string $formatted_line_item The original HTML as a concatenation of <td> elements.
-			 * @param array $line_item_data The relevant data for the line item for which the table row is being generated.
+			 * @param string $line_item_display_data Data to use to generate the HTML table row to be rendered for the line item.
+			 * @param array $line_item_data The relevant data for the line item for which the HTML table row is being generated.
 			 * @param WC_Order $order The order for which the receipt is being generated.
-			 * @return string The actual HTML to put inside the <tr> element corresponding to the line item in the table.
+			 * @return string The actual data to use to generate the HTML for the line item.
 			 *
 			 * @since 8.9.0
 			 */
-			$formatted_line_item    = apply_filters( 'woocommerce_printable_order_receipt_formatted_line_item', $formatted_line_item, $line_item_data, $order );
-			$formatted_line_items[] = $formatted_line_item;
+			$line_item_display_data = apply_filters( 'woocommerce_printable_order_receipt_line_item_display_data', $line_item_display_data, $line_item_data, $order );
+			$attributes             = '';
+			foreach ( $line_item_display_data['tr_attributes'] as $attribute_name => $attribute_value ) {
```

Comment:
@Konamiman Am I correct in thinking that to add a row class you'd filter the `woocommerce_printable_order_receipt_line_item_display_data` value and pass back `$data['tr_attributes']['class'] = 'my-class';` what would be the best way to ensure that Plugin A and Plugin B can both add a class without stomping on each other?

Reply:
And is there a need for escaping/sanitizing the attributes before outputting them?

Reply:
You can check if there's already a `'class'` key in `$data['tr_attributes']`, and if so you add a space and your class name to the existing value. In one line:

```php
$data['tr_attributes']['class'] = trim(($data['tr_attributes']['class'] ?? '') . ' my-class')
```

> And is there a need for escaping/sanitizing the attributes before outputting them?

That's a good point. I've added an `esc_attr` for the attribute values received from the hook.